### PR TITLE
Add stylelint-polaris/media-queries-allowed-list plugin

### PR DESCRIPTION
**Pull request by aaronccasanova**
### WHAT is this pull request doing?

This PR introduces a new `stylelint-polaris/media-queries-allowed-list` plugin to ensure `@media` rules only use the new Polaris `breakpoints` aliases and other support media types.

e.g. `stylelint **/*.scss`

Screen Shot 2022-06-28 at 5 02 04 PM

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

**Review by alex-page — COMMENTED** [review]
Looking at the screenshot, I see the output as errors and I think these should be warnings. We want to encourage consistency and push towards the Polaris values. Once there is a high level of adoption these can become errors. I might be reading this wrong though!

**Comment by aaronccasanova** [comment]
> I see the output as errors and I think these should be warnings.

@alex-page Good call! I [set the internal config to `warn` on `main`](https://github.com/Shopify/polaris/pull/6438/commits/71130808a9adb1ec39fa735e4cd590b5cae0d72c) and will revert it back on the `next-breakpoints` branch.

File: stylelint-polaris/configs/internal.js [hunk 1]
```diff
@@ -38,5 +45,34 @@ module.exports = {
...
+          '$p-breakpoints-xs-up',
+          '$p-breakpoints-xs-down',
+          '$p-breakpoints-xs-only',
+          '$p-breakpoints-sm-up',
+          '$p-breakpoints-sm-down',
+          '$p-breakpoints-sm-only',
+          '$p-breakpoints-md-up',
+          '$p-breakpoints-md-down',
+          '$p-breakpoints-md-only',
+          '$p-breakpoints-lg-up',
```
Comment (sam-b-rose) [inline stylelint-polaris/configs/internal.js]: Are these suppose to be duplicated or should this be `$p-breakpoints-{up,down,only}` for all?

File: stylelint-polaris/index.js [hunk 2]
```diff
@@ -43,5 +50,34 @@ module.exports = {
...
+      {
+        // Allowed media types and media conditions
+        // https://www.w3.org/TR/mediaqueries-5/#media
+        allowedMedia: [
+          'print',
+          'screen',
+          '-ms-high-contrast',
+          'forced-colors',
+          // TODO: Add utility to @shopify/polaris-tokens to getMediaConditionNames
+          '$p-breakpoints-xs-up',
```
Comment (BPScott) [inline stylelint-polaris/index.js]: What if you make it so it accepts a regex, then you wouldn't need to have write out a big list `/^\$p-breakpoints-(xs|sm|md|lg|xl)-(up|down|only)$/`

File: stylelint-polaris/index.js [hunk 3]
```diff
@@ -43,5 +50,34 @@ module.exports = {
...
+          '$p-breakpoints-xs-down',
+          '$p-breakpoints-xs-only',
+          '$p-breakpoints-sm-up',
+          '$p-breakpoints-sm-down',
+          '$p-breakpoints-sm-only',
+          '$p-breakpoints-md-up',
+          '$p-breakpoints-md-down',
+          '$p-breakpoints-md-only',
+          '$p-breakpoints-lg-up',
+          '$p-breakpoints-lg-up',
```
Comment (BPScott) [inline stylelint-polaris/index.js]: $p-breakpoints-lg-up and $p-breakpoints-xl-only are repeated twice while $p-breakpoints-lg-only and $p-breakpoints-xl-up are absent

File: stylelint-polaris/index.js (hunk 2 above)
Comment (aaronccasanova) [reply stylelint-polaris/index.js]: Brilliant https://github.com/Shopify/polaris/pull/6438/commits/6d1f8241e0bd58de5f5fbe4f4c0015a0eb742e61

File: stylelint-polaris/configs/internal.js (hunk 1 above)
Comment (aaronccasanova) [reply stylelint-polaris/configs/internal.js]: Human error, updated to use RegExp here: https://github.com/Shopify/polaris/pull/6438/commits/6d1f8241e0bd58de5f5fbe4f4c0015a0eb742e61

**Review by gui-santos — APPROVED** [review]
🌮

**Review by alex-page — APPROVED** [review]

**Review by lgriffee — APPROVED** [review]
🚢✨🎉

**Comment by aaronccasanova** [comment]
Added a last minute update to the `allowedScssInterpolations` config to [check against the inner expression](https://github.com/Shopify/polaris/pull/6438/commits/e51f0459c94091321fc74787dbda50980573663f). This avoids redundancy by not requiring authors to defined `#{}` to each match pattern e.g.
From `allowedScssInterpolations: ['#{$my-expression}', '#{1 + 1}']`
To `allowedScssInterpolations: ['$my-expression', '1 + 1']`

File: stylelint-polaris/index.js (hunk 3 above)
Comment (aaronccasanova) [reply stylelint-polaris/index.js]: https://github.com/Shopify/polaris/pull/6438#discussion_r910546738

**Pull request merged**